commons-geometry-euclidean/src/main/java/org/apache/commons/geometry/euclidean/twod/hull/ConvexHull2D.java

```diff
@@ -0,0 +1,469 @@
+            points.forEach(this::append);
+            return this;
+        }
+
+        /**
+         * Build a convex hull from the set appended points.
+         *
+         * @return the convex hull
+         * @throws IllegalStateException if generator fails to generate a convex hull for
+         *      the given set of input points
+         */
+        public ConvexHull2D build() {
+            Collection<Vector2D> hullVertices;
+            if (candidates.size() < 2) {
+                hullVertices = candidates;
+            } else {
+                hullVertices = findHullVertices(candidates);
+            }
+
+            if (!isConvex(hullVertices)) {
+                throw new IllegalStateException("Convex hull algorithm failed to generate solution");
+            }
+
+            return new ConvexHull2D(hullVertices, precision);
+        }
+
+        /** Build the convex quadrilateral with the found corner points (with min/max x/y
+         * coordinates).
+         *
+         * @param points the respective points with min/max x/y coordinate
+         * @return the quadrilateral
+         */
+        private static List<Vector2D> buildQuadrilateral(final Vector2D... points) {
+            final List<Vector2D> quadrilateral = new ArrayList<>();
+            for (final Vector2D p : points) {
+                if (!quadrilateral.contains(p)) {
+                    quadrilateral.add(p);
+                }
+            }
+            return quadrilateral;
+        }
+
+        /** Checks if the given point supersedes one of the corners. If it does the old
+         * corner is removed and the point added to the collection of points.
+         *
+         * @param point a given point.
+         */
+        private void checkCorners(Vector2D point) {
+            if (minX == null || point.getX() < minX.getX()) {
+                minX = point;
+                candidates.add(point);
+            }
+            if (maxX == null || point.getX() > maxX.getX()) {
+                maxX = point;
+                candidates.add(point);
+            }
+            if (minY == null || point.getY() < minY.getY()) {
+                minY = point;
+                candidates.add(point);
+            }
+            if (maxY == null || point.getY() > maxY.getY()) {
+                maxY = point;
+                candidates.add(point);
+            }
+        }
+
+        /** Checks if the given point is located within the convex quadrilateral.
+         * @param point the point to check
+         * @param quadrilateralPoints the convex quadrilateral, represented by 4 points
+         * @return {@code true} if the point is inside the quadrilateral, {@code false} otherwise
+         */
+        private boolean insideQuadrilateral(final Vector2D point,
+                                                   final List<? extends Vector2D> quadrilateralPoints) {
+
+            Vector2D v1 = quadrilateralPoints.get(0);
```

Comment:
Since this walks a circle around the quadrilateral we can start anywhere. IIUC this can be changed to:
```java
final int size = quadrilateralPoints.size();
Vector2D v1 = quadrilateralPoints.get(size - 1);
Vector2D v2 = quadrilateralPoints.get(0);

// ...

for (int i = 1; i < size; i++) {
    v1 = v2;
    v2 = quadrilateralPoints.get(i);
```

Reply:
Yes, the change could be made here. More importantly it showed that I missed an edge case when collinear points should be included. I fixed that as well.